server/balancer.go

```diff
@@ -0,0 +1,438 @@
+	followerPeers, excludedStores := getFollowerPeers(region, leader)
+	newLeader := rb.selectNewLeaderPeer(cluster, followerPeers)
+	if newLeader == nil {
+		log.Warn("new leader peer cannot be found to do balance, try to do follower peer balance")
+		return nil, false, nil
+	}
+
+	// Select one store to add new peer.
+	newPeer, err := rb.selectAddPeer(cluster, stores, excludedStores)
+	if err != nil {
+		return nil, false, errors.Trace(err)
+	}
+	if newPeer == nil {
+		log.Warn("new peer cannot be found to do balance")
+		return nil, true, nil
+	}
+
+	if !rb.checkScore(cluster, leader, newPeer) {
+		return nil, false, nil
+	}
+
+	leaderTransferOperator := newTransferLeaderOperator(leader, newLeader, maxWaitCount)
+	addPeerOperator := newAddPeerOperator(newPeer)
+	removePeerOperator := newRemovePeerOperator(leader)
+
+	return newBalanceOperator(region, leaderTransferOperator, addPeerOperator, removePeerOperator), true, nil
+}
+
+func (rb *resourceBalancer) doFollowerBalance(cluster *clusterInfo, stores []*storeInfo, region *metapb.Region, leader *metapb.Peer) (*balanceOperator, error) {
+	followerPeers, excludedStores := getFollowerPeers(region, leader)
+
+	newPeer, err := rb.selectAddPeer(cluster, stores, excludedStores)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if newPeer == nil {
+		log.Warn("new peer cannot be found to do balance")
+		return nil, nil
+	}
+
+	oldPeer, err := rb.selectRemovePeer(cluster, followerPeers)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if oldPeer == nil {
+		log.Warnf("find no store to remove peer for region %v", region)
+		return nil, nil
+	}
+
+	if !rb.checkScore(cluster, oldPeer, newPeer) {
+		return nil, nil
+	}
+
+	addPeerOperator := newAddPeerOperator(newPeer)
+	removePeerOperator := newRemovePeerOperator(oldPeer)
+	return newBalanceOperator(region, addPeerOperator, removePeerOperator), nil
+}
+
+func (rb *resourceBalancer) Balance(cluster *clusterInfo) (*balanceOperator, error) {
+	stores := cluster.getStores()
+	region, oldLeader, isLeaderBalance := rb.selectBalanceRegion(cluster, stores)
+	if region == nil || oldLeader == nil {
+		log.Warn("region cannot be found to do balance")
+		return nil, nil
+	}
+
+	// If region peer count is not equal to max peer count, no need to do capacity balance.
+	if len(region.GetPeers()) != int(cluster.getMeta().GetMaxPeerCount()) {
+		log.Warnf("region peer count %d not equals to max peer count %d, no need to do balance",
+			len(region.GetPeers()), cluster.getMeta().GetMaxPeerCount())
+		return nil, nil
+	}
+
+	if isLeaderBalance {
+		ops, ok, err := rb.doLeaderBalance(cluster, stores, region, oldLeader)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if !ok {
+			// If we cannot find a region peer to do leader transfer,
+			// then we do follower balance instead.
+			return rb.doFollowerBalance(cluster, stores, region, oldLeader)
```

Comment:
Why should we balance follower peer? The follower must not be on the Store we selected before.